File: TwoPointers/LinkedList/Leetcode19.java
public class Leetcode19 {
    //Leetcode 19
    /**
     * This problem want to the remove the nth node from its end
     */
    public ListNode removeNthFromEnd(ListNode head, int n) {
        if(head == null)return null;
        ListNode dummy = new ListNode(-1);
        dummy.next = head;
        ListNode find = findKth(dummy,n+1);
        find.next = find.next.next;
        return dummy.next;
    }
    
    // Let first use this helper function to find the n+1 node fromt the end.
    // It is using fast-slow pointer. Fast point will move k times first, then slow start to move.
    // When fast hit the end, the slow is exactlly what we want.
    public ListNode findKth(ListNode head, int k){
        if(head == null)return null;
        ListNode fast = head;
        for(int i = 0;i< k;i++){
            fast = fast.next;
        }
        ListNode slow = head;
        while(fast != null){
            slow = slow.next;
            fast = fast.next;
        }
        return slow;
    }
}

File: LinkedList/Python/Leetcode19.py
class Leetcode19:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummy = ListNode()
        dummy.next = head
        left, right = dummy, head
        for i in range(n):
            right = right.next
        while right:
            left = left.next
            right = right.next
        left.next = left.next.next
        return dummy.next
